parquet-column/src/main/java/parquet/column/values/delta/DeltaBinaryPackingValuesWriterForInteger.java

```diff
@@ -0,0 +1,232 @@
+ * <p/>
+ * <pre>
+ *   {@code
+ *     delta-binary-packing: <page-header> <block>*
+ *     page-header := <block size in values> <number of miniblocks in a block> <total value count> <first value>
+ *     block := <min delta> <list of bitwidths of miniblocks> <miniblocks>
+ *
+ *     min delta : zig-zag var int encoded
+ *     bitWidthsOfMiniBlock : 1 byte little endian
+ *     blockSizeInValues,blockSizeInValues,totalValueCount,firstValue : unsigned varint
+ *   }
+ * </pre>
+ *
+ * The algorithm and format is inspired by D. Lemire's paper: http://lemire.me/blog/archives/2012/09/12/fast-integer-compression-decoding-billions-of-integers-per-second/
+ *
+ * @author Tianshuo Deng
+ */
+public class DeltaBinaryPackingValuesWriterForInteger {
+  /**
+   * max bitwidth for a mini block, it is used to allocate miniBlockByteBuffer which is
+   * reused between flushes.
+   */
+  public static final int MAX_BITWIDTH = 32;
+
+  /**
+   * stores blockSizeInValues, miniBlockNumInABlock and miniBlockSizeInValues
+   */
+  private final DeltaBinaryPackingConfig config;
+
+  /**
+   * bit width for each mini block, reused between flushes
+   */
+  private final int[] bitWidths;
+
+  private int totalValueCount = 0;
+
+  /**
+   * a pointer to deltaBlockBuffer indicating the end of deltaBlockBuffer
+   * the number of values in the deltaBlockBuffer that haven't flushed to baos
+   * it will be reset after each flush
+   */
+  private int deltaValuesToFlush = 0;
+
+  /**
+   * stores delta values starting from the 2nd value written(1st value is stored in header).
+   * It's reused between flushes
+   */
+  private int[] deltaBlockBuffer;
+
+  /**
+   * bytes buffer for a mini block, it is reused for each mini block.
+   * Therefore the size of biggest miniblock with bitwith of MAX_BITWITH is allocated
+   */
+  private byte[] miniBlockByteBuffer;
+
+  /**
+   * firstValue is written to the header of the page
+   */
+  private int firstValue = 0;
+
+  /**
+   * cache previous written value for calculating delta
+   */
+  private int previousValue = 0;
+
+  /**
+   * min delta is written to the beginning of each block.
+   * it's zig-zag encoded. The deltas stored in each block is actually the difference to min delta,
+   * therefore are all positive
+   * it will be reset after each flush
+   */
+  private int minDeltaInCurrentBlock = Integer.MAX_VALUE;
+
+  public DeltaBinaryPackingValuesWriterForInteger(final DeltaBinaryPackingConfig config) {
+    this.config = config;
+    bitWidths = new int[config.miniBlockNumInABlock];
+    deltaBlockBuffer = new int[config.blockSizeInValues];
+    miniBlockByteBuffer = new byte[config.miniBlockSizeInValues * MAX_BITWIDTH];
+  }
+
+  public void writeInteger(int v, final CapacityByteArrayOutputStream baos) {
+    totalValueCount++;
+
+    if (totalValueCount == 1) {
+      firstValue = v;
+      previousValue = firstValue;
+      return;
+    }
+
+    int delta = v - previousValue;//calculate delta
```

Comment:
OK, I will add two comments. One for this overflow and one for the INT32 encoding (that should not be done by the INT64 encoder). 

Yes, we have unit tests for [the INT32 case](https://github.com/lunchev/incubator-parquet-mr/commit/bd6d1b398c0e4b43355ae112743500792bdf6edd#diff-7bc3f0a262f0aac6518722769ebcd270R129) and for [the INT64 case](https://github.com/lunchev/incubator-parquet-mr/commit/bd6d1b398c0e4b43355ae112743500792bdf6edd#diff-7bc3f0a262f0aac6518722769ebcd270L142).  I have walked through those two tests with a debugger to check exactly how they works and yes, I think that they check the works case scenario for the overflow. We also have [randomDataTest()](https://github.com/lunchev/incubator-parquet-mr/blob/bd6d1b398c0e4b43355ae112743500792bdf6edd/parquet-column/src/test/java/parquet/column/values/delta/DeltaBinaryPackingValuesWriterTest.java#L241) and [randomDataTestLong()](https://github.com/lunchev/incubator-parquet-mr/blob/bd6d1b398c0e4b43355ae112743500792bdf6edd/parquet-column/src/test/java/parquet/column/values/delta/DeltaBinaryPackingValuesWriterTest.java#L225) that also hit quite a lot of overflowing cases and they still work perfectly fine.

@julienledem Do you need me to check some other scenario?